File: clients/android/app/src/main/java/live/yunmao/app/network/WSClient.kt
package live.yunmao.app.network

import io.ktor.client.HttpClient
import io.ktor.client.plugins.websocket.WebSockets
import io.ktor.client.plugins.websocket.wss
import io.ktor.websocket.Frame
import io.ktor.websocket.WebSocketSession
import io.ktor.websocket.readText
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

sealed class WSEvent {
    data class FeedStateChange(val roomId: String, val payload: JsonObject) : WSEvent()
    data class ChatMessage(val data: JsonObject) : WSEvent()
    data class ChatModeration(val messageId: String, val action: String) : WSEvent()
    data class Reconnected(val attempt: Int) : WSEvent()
    data class Disconnected(val cause: Throwable?) : WSEvent()
    data class Unknown(val type: String, val payload: JsonObject) : WSEvent()
}

/**
 * Android WSClient with exponential-backoff auto-reconnect.
 *
 * Mirrors the iOS `WSClient` (see `clients/ios/.../Network/WSClient.swift`).
 * Subscribed room list is preserved across reconnects so the UI does not need to
 * re-issue subscribe calls when the socket is re-established after a network churn
 * or gateway failover.
 */
class WSClient(
    private val wsBase: String,
    private val token: String,
    var autoReconnect: Boolean = true,
    var baseBackoffMs: Long = 1_000L,
    var maxBackoffMs: Long = 30_000L,
    var maxAttempts: Int = 5,
) {
    private val events = Channel<WSEvent>(Channel.BUFFERED)
    fun events(): Flow<WSEvent> = events.receiveAsFlow()

    @Volatile private var session: WebSocketSession? = null
    private val subscribedRooms = LinkedHashSet<String>()
    private var connectJob: Job? = null

    suspend fun connect(client: HttpClient) {
        val urlNoScheme = wsBase.removePrefix("wss://").removePrefix("ws://")
        client.wss(host = urlNoScheme, path = "/ws") {
            session = this
            outgoing.send(Frame.Text("""{"type":"auth","token":"$token"}"""))
            for (frame in incoming) {
                val text = (frame as? Frame.Text)?.readText() ?: continue
                handle(text)
            }
        }
    }

    /**
     * Wraps [connect] with automatic retry on disconnect, using exponential backoff
     * (1s, 2s, 4s, 8s, 16s, … capped at [maxBackoffMs]), up to [maxAttempts] tries.
     * After each successful reconnect the cached [subscribedRooms] are re-subscribed.
     */
    suspend fun connectWithRetry(client: HttpClient) {
        var attempt = 0
        var backoff = baseBackoffMs
        while (true) {
            try {
                connect(client)
                // Re-establish subscriptions on a clean connect/reconnect.
                val rooms = synchronized(subscribedRooms) { subscribedRooms.toList() }
                for (r in rooms) {
                    session?.outgoing?.send(
                        Frame.Text("""{"type":"subscribe","room_id":"$r"}""")
                    )
                }
                if (attempt > 0) {
                    events.send(WSEvent.Reconnected(attempt))
                }
                return
            } catch (ce: CancellationException) {
                throw ce
            } catch (t: Throwable) {
                attempt++
                events.send(WSEvent.Disconnected(t))
                if (!autoReconnect || attempt >= maxAttempts) {
                    return
                }
                delay(backoff)
                backoff = (backoff * 2).coerceAtMost(maxBackoffMs)
            }
        }
    }

    suspend fun subscribe(roomId: String) {
        synchronized(subscribedRooms) { subscribedRooms.add(roomId) }
        session?.outgoing?.send(Frame.Text("""{"type":"subscribe","room_id":"$roomId"}"""))
    }

    suspend fun sendChat(roomId: String, body: String, clientId: String) {
        // 用 Json 库正确转义 body（含引号、控制字符），避免手拼 JSON 字符串。
        val payload = buildString {
            append("{\"type\":\"chat.send\",\"room_id\":\"")
            append(roomId)
            append("\",\"body\":")
            append(Json.encodeToString(kotlinx.serialization.builtins.serializer(), body))
            append(",\"client_msg_id\":\"")
            append(clientId)
            append("\"}")
        }
        session?.outgoing?.send(Frame.Text(payload))
    }

    private suspend fun handle(text: String) {
        runCatching {
            val obj = Json.parseToJsonElement(text).jsonObject
            val type = obj["type"]?.jsonPrimitive?.contentOrNull ?: "unknown"
            val ev: WSEvent = when (type) {
                "feed.state_change", "feed.created", "feed.completed" -> {
                    val roomId = obj["room_id"]?.jsonPrimitive?.contentOrNull ?: ""
                    WSEvent.FeedStateChange(roomId, obj)
                }
                "room.chat.message", "chat.message" -> WSEvent.ChatMessage(obj)
                "room.chat.moderation", "chat.moderation" -> {
                    val mid = obj["message_id"]?.jsonPrimitive?.contentOrNull ?: ""
                    val act = obj["action"]?.jsonPrimitive?.contentOrNull ?: "hide"
                    WSEvent.ChatModeration(mid, act)
                }
                else -> WSEvent.Unknown(type, obj)
            }
            events.send(ev)
        }
    }
}

File: clients/ios/YunmaoApp/Sources/YunmaoApp/Network/WSClient.swift
// WSClient：URLSessionWebSocketTask 封装，处理订阅房间 + feed.* / chat.* / room.chat.moderation
// 事件。事件通过 AsyncStream 暴露。
//
// Phase 10 增强：自动指数退避重连（`reconnect` / `autoReconnect`），覆盖网络抖动、
// 前后台切换场景。`subscribedRooms` 缓存订阅列表，重连后自动恢复订阅。

import Foundation

public enum WSEvent {
    case feedStateChange(roomId: String, payload: [String: Any])
    case chatMessage(ChatMessage)
    case chatModeration(messageId: String, action: String)
    case reconnected
    case disconnected(Error?)
    case unknown(String, [String: Any])
}

public actor WSClient {
    private let url: URL
    private var task: URLSessionWebSocketTask?
    private var token: String
    private var continuation: AsyncStream<WSEvent>.Continuation?
    private var subscribedRooms: [String] = []
    private var reconnecting = false
    public var autoReconnect: Bool
    public var baseBackoffSeconds: Double
    public var maxBackoffSeconds: Double

    public init(url: URL, token: String, autoReconnect: Bool = true,
                baseBackoffSeconds: Double = 1.0, maxBackoffSeconds: Double = 30.0) {
        self.url = url
        self.token = token
        self.autoReconnect = autoReconnect
        self.baseBackoffSeconds = baseBackoffSeconds
        self.maxBackoffSeconds = maxBackoffSeconds
    }

    public func events() -> AsyncStream<WSEvent> {
        AsyncStream { cont in
            self.continuation = cont
            cont.onTermination = { @Sendable _ in
                Task { await self.close() }
            }
        }
    }

    public func connect() async {
        var req = URLRequest(url: url)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        task = URLSession.shared.webSocketTask(with: req)
        task?.resume()
        Task { await pump() }
    }

    public func reconnect() async {
        guard !reconnecting else { return }
        reconnecting = true
        defer { reconnecting = false }
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        var delay = baseBackoffSeconds
        for attempt in 0..<5 {
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            await connect()
            let rooms = subscribedRooms
            for roomId in rooms {
                await subscribe(roomId: roomId)
            }
            continuation?.yield(.reconnected)
            return
        }
    }

    public func subscribe(roomId: String) async {
        if !subscribedRooms.contains(roomId) {
            subscribedRooms.append(roomId)
        }
        let msg: [String: Any] = ["type": "subscribe", "room_id": roomId]
        try? await sendJSON(msg)
    }

    public func sendChat(roomId: String, body: String, clientID: String) async throws {
        let msg: [String: Any] = [
            "type": "chat.send",
            "room_id": roomId,
            "body": body,
            "client_msg_id": clientID,
        ]
        try await sendJSON(msg)
    }

    public func close() async {
        autoReconnect = false
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        continuation?.finish()
    }

    private func sendJSON(_ msg: [String: Any]) async throws {
        guard let task else { return }
        let data = try JSONSerialization.data(withJSONObject: msg)
        try await task.send(.data(data))
    }

    private func pump() async {
        guard let task else { return }
        while task.state == .running {
            do {
                let m = try await task.receive()
                switch m {
                case .string(let s):
                    handle(s.data(using: .utf8) ?? Data())
                case .data(let d):
                    handle(d)
                @unknown default: break
                }
            } catch {
                continuation?.yield(.disconnected(error))
                if autoReconnect {
                    Task { await self.reconnect() }
                } else {
                    continuation?.finish()
                }
                return
            }
        }
    }

    private func handle(_ data: Data) {
        guard let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }
        let type = (dict["type"] as? String) ?? "unknown"
        switch type {
        case "feed.state_change", "feed.created", "feed.completed":
            let roomId = (dict["room_id"] as? String) ?? ""
            continuation?.yield(.feedStateChange(roomId: roomId, payload: dict))
        case "room.chat.message", "chat.message":
            if let payload = dict["payload"] as? [String: Any],
               let json = try? JSONSerialization.data(withJSONObject: payload),
               let msg = try? JSONDecoder.yunmao.decode(ChatMessage.self, from: json) {
                continuation?.yield(.chatMessage(msg))
            }
        case "room.chat.moderation", "chat.moderation":
            let mid = (dict["message_id"] as? String) ?? ""
            let action = (dict["action"] as? String) ?? "hide"
            continuation?.yield(.chatModeration(messageId: mid, action: action))
        default:
            continuation?.yield(.unknown(type, dict))
        }
    }
}
